File: sections/65_polynomial_algorithms_recurrences/problems/c_huge_linear_recurrence/solution.py
import sys

MOD = 998_244_353


def combine(left: list[int], right: list[int], coefficient: list[int]) -> list[int]:
    order = len(coefficient)
    product = [0] * (2 * order - 1)
    for i, a in enumerate(left):
        for j, b in enumerate(right):
            product[i + j] = (product[i + j] + a * b) % MOD
    for degree in range(2 * order - 2, order - 1, -1):
        value = product[degree]
        for back, coefficient_value in enumerate(coefficient, 1):
            product[degree - back] = (
                product[degree - back] + value * coefficient_value
            ) % MOD
    return product[:order]


def nth_term(initial: list[int], coefficient: list[int], index: int) -> int:
    order = len(initial)
    if index < order:
        return initial[index]
    result = [1] + [0] * (order - 1)
    base = [coefficient[0]] if order == 1 else [0, 1] + [0] * (order - 2)
    while index:
        if index & 1:
            result = combine(result, base, coefficient)
        base = combine(base, base, coefficient)
        index >>= 1
    return sum(a * b for a, b in zip(result, initial)) % MOD


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    order, index = data[:2]
    initial = data[2 : 2 + order]
    coefficient = data[2 + order : 2 + 2 * order]
    print(nth_term(initial, coefficient, index))


if __name__ == "__main__":
    main()

File: sections/65_polynomial_algorithms_recurrences/problems/c_huge_linear_recurrence/solution.cpp
#include <bits/stdc++.h>
using namespace std;

using int64 = long long;
constexpr int64 MOD = 998244353;

vector<int64> combine(
    const vector<int64> &left,
    const vector<int64> &right,
    const vector<int64> &coefficient
) {
    int order = coefficient.size();
    vector<int64> product(2 * order - 1);
    for (int i = 0; i < order; ++i)
        for (int j = 0; j < order; ++j)
            product[i + j] = (product[i + j] + left[i] * right[j]) % MOD;
    for (int degree = 2 * order - 2; degree >= order; --degree)
        for (int back = 1; back <= order; ++back)
            product[degree - back] = (
                product[degree - back] + product[degree] * coefficient[back - 1]
            ) % MOD;
    product.resize(order);
    return product;
}

int64 nth_term(
    const vector<int64> &initial,
    const vector<int64> &coefficient,
    unsigned long long index
) {
    int order = initial.size();
    if (index < static_cast<unsigned long long>(order)) return initial[index];
    vector<int64> result(order), base(order);
    result[0] = 1;
    if (order == 1)
        base[0] = coefficient[0];
    else
        base[1] = 1;
    while (index > 0) {
        if (index & 1) result = combine(result, base, coefficient);
        base = combine(base, base, coefficient);
        index >>= 1;
    }
    int64 answer = 0;
    for (int i = 0; i < order; ++i)
        answer = (answer + result[i] * initial[i]) % MOD;
    return answer;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int order;
    unsigned long long index;
    if (!(cin >> order >> index)) return 0;
    vector<int64> initial(order), coefficient(order);
    for (int64 &value : initial) cin >> value;
    for (int64 &value : coefficient) cin >> value;
    cout << nth_term(initial, coefficient, index) << '\n';
}
